bento4 1.5.0-615 (new formula)

slhck opened the pull request:
Adds a formula for the bento4 set of tools to (re)mux or edit MP4
file.

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

Please bear with me – this is my first formula and it's taken me long enough to figure this out. Some things that I haven't been able to consider due to my lack of experience with the program, build tools and Homebrew in particular:

- I had to use the standard `make` way to build these tools (from the `any-gnu-gcc` directory), although it supplies Cmake and a macOS-specific configuration – but I couldn't figure out how to build with this.
- I did not copy any header files or libraries, although this might be useful for programmatic use of the library.
- The formula provides `mp42ts` whereas `gpac` provides `MP4TS` (capitals). I guess this conflicts, so I excluded it.
- I only considered the release `.tar.gz`, although I think building from source would be possible

fxcoudert commented on Formula/bento4.rb:
```diff
@@ -0,0 +1,28 @@
+  conflicts_with "gpac", :because => "both install `mp42ts` binaries"
+
+  def install
+    binaries = %w(aac2mp4 avcinfo hevcinfo mp42aac mp42hevc mp42hls mp42ts
```
As a matter of style, we prefer `%w[` for the array, and indentation of the next line so that subsequent items are all left-aligned after the `[`.

ilovezfs replied:
Does
```
bin.install Dir["*"].select { |f| File.executable? f }
```
install things it shouldn't?

ilovezfs replied:
And if that pulls in things it shouldn't, then
```
bin.install Dir["*"].select { |f| File.file?(f) && File.executable?(f) }
```
may do the right thing.

slhck replied:
Yes, this works better, thanks.

fxcoudert commented on Formula/bento4.rb:
```diff
@@ -0,0 +1,28 @@
+    mp4compact mp4dcfpackager mp4decrypt mp4dump mp4edit mp4encrypt mp4extract
+    mp4fragment mp4info mp4mux mp4rtphintinfo mp4split mp4tag)
+
+    Dir.chdir "Build/Targets/any-gnu-gcc"
```
Here let's use the block version: `chdir "Build/Targets/any-gnu-gcc" do`

fxcoudert commented on Formula/bento4.rb:
```diff
@@ -0,0 +1,28 @@
+    Dir.chdir "Build/Targets/any-gnu-gcc"
+    system "make", "AP4_BUILD_CONFIG=Release"
+
+    Dir.chdir "Release"
```
Let's not `chdir` here and then do `bin.install Release/binary_name`

slhck replied:
How exactly? `bin.install Release/binary_name` is not valid syntax … should I do `bin.install "Release/#{binary_name}"` instead?

When I do that, I get:

    Error: No such file or directory - Release/aac2mp4

slhck replied:
Ah, never mind, it obviously had to be inside the `chdir … do` block. Now it would look like this:

    chdir "Build/Targets/any-gnu-gcc" do
      system "make", "AP4_BUILD_CONFIG=Release"
      binaries.each do |binary_name|
        bin.install "Release/#{binary_name}"
      end
    end

Is that acceptable?

fxcoudert replied:
Yes that's good — clear and concise

ilovezfs replied:
Should be `cd` not `chdir`

fxcoudert commented on Formula/bento4.rb:
```diff
@@ -0,0 +1,28 @@
+  end
+
+  test do
+    system "#{bin}/mp4mux", "--track", test_fixtures("test.m4a"), "#{testpath}/out.mp4"
```
`testpath/"out.mp4"` instead of the `#`

ilovezfs replied:
does just `out.mp4` not work?

slhck replied:
I guess, yeah, I had copied that syntax from here so I assumed it was okay: https://github.com/Homebrew/homebrew-core/blob/master/Formula/gpac.rb

fxcoudert commented on Formula/bento4.rb:
```diff
@@ -0,0 +1,28 @@
+
+  test do
+    system "#{bin}/mp4mux", "--track", test_fixtures("test.m4a"), "#{testpath}/out.mp4"
+    File.exist? "#{testpath}/out.mp4"
```
`File.exist? "out.mp4"` should suffice, we're already in that directory

ilovezfs replied:
should use `assert_predicate`

ilovezfs commented on Formula/bento4.rb:
```diff
@@ -0,0 +1,28 @@
+
+    Dir.chdir "Release"
+    binaries.each do |binary_name|
+      bin.install binary_name
```
A `make install` target needs to be requested upstream.

slhck replied:
Just for my understanding: does that block this PR?

ilovezfs replied:
The request needs to be made with a link to it here.

slhck replied:
Hm, then I don't know what to request, exactly, because I'm not an expert on how that `make install` target should look like or how to communicate it to the project, in terms of what Homebrew needs them to do. I will have a look, but I can't promise that this will ever be fulfilled.

(And just for my understanding, why would this “simple” kind of installation not work, when other formulae also sometimes use `bin.install`?)

ilovezfs replied:
>Hm, then I don't know what to request

Typically it will be either
```
./configure --prefix=/foo/bar
make install
```
or
```
make PREFIX=/foo/bar install
```

>(And just for my understanding, why would this “simple” kind of installation not work, when other formulae also sometimes use bin.install?)

We want the logic maintained upstream not in the formula.

slhck replied:
I see @fxcoudert has already taken care of this: https://github.com/axiomatic-systems/Bento4/issues/165 — thanks a lot!

slhck commented:
Er. I guess I screwed up pushing this branch. What should I do?

ilovezfs commented:
what's the output of
```
git remote
```

slhck commented:
I tried to squash my local commits (from `/usr/local/Homebrew/Library/Taps/homebrew/homebrew-core`) in my separate branch, then tried to push that branch to my fork, but that got rejected. So I pulled the branch from my fork (`git pull https://github.com/slhck/homebrew-core.git add_bento4`), fixed the conflicts, and pushed to that.

The remote of `/usr/local/Homebrew/Library/Taps/homebrew/homebrew-core` is `origin` pointing to `//github.com/Homebrew/homebrew-core`.

ilovezfs commented:
```
cd /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core
mv Formula/bento4.rb ~/
git remote add slhck https://github.com/slhck/homebrew-core.git
git fetch --all
git reset --hard origin/master
git checkout add_bento4
git reset --hard origin/master
cp ~/bento4.rb Formula
git add Formula/bento4.rb
git commit -m "bento4 1.5.0-615 (new formula)"
git push -f slhck add_bento4
```

slhck commented:
You're a lifesaver, thank you. I think I've addressed all comments, so we're just going to have to wait for https://github.com/axiomatic-systems/Bento4/issues/165 — thanks for the reviews and comments.

ilovezfs commented on Formula/bento4.rb:
```diff
@@ -0,0 +1,21 @@
+
+  test do
+    system "#{bin}/mp4mux", "--track", test_fixtures("test.m4a"), "out.mp4"
+    assert_predicate File.exist? "out.mp4"
```
```
    assert_predicate testpath/"out.mp4", :exist?, "Failed to create out.mp4!"
```

slhck replied:
Ah, right. Is it okay if I edit my local file, then do:

```
git add Formula/bento4.rb
git commit --amend -m 'bento4 1.5.0-615 (new formula)'
git push -f slhck add_bento4
```

So everything is in one commit?

ilovezfs replied:
Yeah, but you just want `git commit --amend --no-edit`

slhck replied:
Ok, done, thanks. (Should this all be part of the Formula Cookbook? It's nicely described how to get to your first PR but once you have to do a change to your formula it's easy to 1) not know what to do, 2) screw up and 3) not remember what you did.)

ilovezfs commented:
Thanks for the :new: formula @slhck! :firew

slhck commented:
Great, thanks for your support, happy to contribute to Homebrew!
